src/features/device-state/routes/state.ts

```diff
@@ -25,6 +26,161 @@ export const varListInsert = (varList: EnvVarList, obj: Dictionary<string>) => {
 	});
 };
 
+export type App = {
+	name: string;
+	commit: string;
+	releaseId: number;
+	appId?: string;
+	uuid?: string;
+	services: {
+		[id: number]: AppService;
+	};
+	volumes: AnyObject;
+	networks: AnyObject;
+};
+
+export type AppService = {
+	imageId: number;
+	serviceName: string;
+	image: string;
+	running: boolean;
+	environment: Dictionary<string>;
+	labels: Dictionary<string>;
+	contract?: string;
+};
+
+export type LocalState = {
+	name: string;
+	config: Dictionary<string>;
+	apps: Dictionary<Partial<App>>;
+};
+
+async function buildAppFromRelease(
+	device: AnyObject,
+	application: AnyObject,
+	release: AnyObject,
+	config: Dictionary<string>,
+): Promise<App> {
+	let composition: AnyObject = {};
+	const services: App['services'] = {};
+
+	// Parse the composition to forward values to the device
+	if (_.isObject(release.composition)) {
+		composition = release.composition;
+	} else {
+		try {
+			composition = JSON.parse(release.composition);
+		} catch {
+			composition = {};
+		}
+	}
+
+	(release.contains__image as AnyObject[]).forEach((ipr) => {
+		// extract the per-image information
+		const image = ipr.image[0];
+
+		const serviceInstall = (() => {
+			const si = serviceInstallFromImage(device, image);
+
+			if (si != null) {
+				return si;
+			}
+
+			if (image.is_a_build_of__service[0] == null) {
+				throw new Error(
+					`Could not find service install for device: '${
+						device.uuid
+					}', image: '${image?.id}', service: '${JSON.stringify(
+						image?.is_a_build_of__service,
+					)}', service installs: '${JSON.stringify(device.service_install)}'`,
+				);
+			}
+
+			/**
+			 * Return a fake service install as this is what we will use for MultiApp
+			 */
+			return {
+				service: [image.is_a_build_of__service[0]],
+				device_service_environment_variable: [],
+			};
+		})();
```

Comment:
This looks like functional changes rather than pure refactoring?

Reply:
removed